src/trio/_tests/test_exports.py

```diff
@@ -572,3 +573,72 @@ def test_classes_are_final() -> None:
                 continue
 
             assert class_is_final(class_)
+
+
+def test_pyright_recognizes_init_attributes() -> None:
+    # Obviously, this isn't completely accurate.
+    # It should still be good enough. Hopefully.
+    # (attrs updates fields before we can access them)
+    files = []
+
+    for path in (Path(inspect.getfile(trio)) / "..").glob("**/*.py"):
+        with open(path, "rb") as f:
+            files.append(list(tokenize.tokenize(f.readline)))
+
+    for module in PUBLIC_MODULES:
+        for name, class_ in module.__dict__.items():
+            if not attrs.has(class_):
+                continue
+            if isinstance(class_, _util.NoPublicConstructor):
+                continue
+
+            file = None
+            start = None
+            for contents in files:
+                last_was_class = False
+                for i, token in enumerate(contents):
+                    if (
+                        token.type == tokenize.NAME
+                        and token.string == name
+                        and last_was_class
+                    ):
+                        assert file is None
+                        file = contents
+                        start = i - 1
+
+                    if token.type == tokenize.NAME and token.string == "class":
+                        last_was_class = True
+                    else:
+                        last_was_class = False
+
+            assert file is not None, f"{name}: {class_!r}"
```

Comment:
I think you might also want to `assert start is not None`, so the below typecheck succeeds.